### DISCUSSION TITLE
[Ps1 WebAssembly project] Make powershell v9 great again

### DISCUSSION THREAD

**Discussion by kasini3000** [PowerShell/PowerShell#17339] [Enhancements]
A few years ago, nginx had a multilingual dynamic web server, which I suggested to support powershell, but the suggestion was not accepted. The nginx team thought this was incredible.

A few years ago, I suggested that the ps team support the emoj large character set, but the suggestion was not accepted. Maybe the PS team thinks it shouldn't do this. But win terminal did it and it worked.

Now, I heard the news, python is going to be embedded in the browser via WebAssembly! I think it's time to call powershell enthusiasts to do the ps1 embedded browser thing through WebAssembly. I think this is a big deal for powershell. If we do nothing, powershell will inevitably be sidelined (decline) In the future.

--------

news：

Anaconda developers recently released PyScript, a framework that lets developers mix Python code in HTML, which some call "JSP for Python". In fact, PyScript uses WebAssembly under the hood because it is built on Pyodide, which consists of a CPython 3.8 interpreter compiled to WebAssembly, allowing Python to be run in a web browser.

![144227_dh3S_2720166](https://user-images.githubusercontent.com/65390418/168341665-f18f0b80-a8e2-4a7d-9e19-4ad316160282.png)

**Comment by Ayanmullick** [comment]
<details>

<summary>Repl has a PowerShell template. </summary>

<a href="https://replit.com/@Ayanmullick/PowerShell-template#main.ps1">
         <img alt="Qries" src="https://user-images.githubusercontent.com/17693249/168359082-572d1d5c-ff38-4983-aba7-518858dd4ea4.png">
      </a>
</details>


However, the output works on the browser console, not on the webpage.
@adamdriscoll

**Comment by Ayanmullick** [comment]
<a href="https://www.youtube.com/embed/lP_qdhAHFlg?start=2712&end=2765" title="1. Click to see video➩.NET on WASI NDC Porto 2022" target="_blank"><img align="right" src="http://img.youtube.com/vi/lP_qdhAHFlg/0.jpg" width="180" height="120" border="10"/></a>

1.  @SteveSanderson's demo of a .NET 7 console app running on the browser using WASI.

2. Azure data plane cmdlets consuming cloud APIs might not need a hard dependency on the host file system, as explained by Lin Clark from the [Bytecode Alliance](https://bytecodealliance.org/), @dingmeng-xue


@StartAutomating, regarding the use case...

<a href="https://www.youtube.com/embed/k3HDivg3xyc?start=263&end=606" title="2. Click to see video➩Keynote: WASI" target="_blank"><img align="right" src="http://img.youtube.com/vi/k3HDivg3xyc/1.jpg" width="180" height="120" border="10"/></a>

- For example, if one wants to [hibernate an Azure VM](https://github.com/Azure/azure-cli/issues/19611) one needs functions, other resources etc. vs. one could just create a shortcut like below and add to the Azure portal.
https://winaero.com/create-shortcut-ps1-powershell-file-windows-10

- One could [run Read-SQLTableData right on the browser to fetch AzSQL data on a static site](https://github.com/microsoft/SQLServerPSModule/issues/39) and display a report without needing any middleware.

**Comment by rhubarb-geek-nz** [comment]
> I think this is a big deal for powershell. If we do nothing, powershell will inevitably be sidelined (decline) In the future.

I don't think that is the case. PowerShell Core is built on DotNET Core, the best way to use DotNET Core in websites is through ASP.NET.

PowerShell is not a general purpose programming language like Python. PowerShell sits much higher on the stack and is scripting language to orchestrate other components. PowerShell has plenty to do in tasks for which it was intended such as managing servers. For instance this is where it shines, I can now write cross platform scripts to manage complex tasks for administering docker containers for example. I would prefer it does that job well rather than trying to become something to write webpages in.

**Reply by kasini3000** [reply]
@rhubarb-geek-nz  see my powershell automation project (in my repo)

**Reply by kasini3000** [reply]
Other users don't want to see more about my project HERE, so I'm very sorry. You are welcome to come here for more discussions,   https://github.com/kasini3000/kasini3000/issues

**Comment by StartAutomating** [comment]
A few thoughts:

1. From a purely practical point of view, it seems as if nothing short of raw WebAssembly is actually written in WebAssembly.  Everything else is just compiled down to WebAssembly.

This means the question isn't really so much "should PowerShell work as WebAssembly" (engine and all), but "how would we compile PowerShell into a WebAssembly".

I see value in having the whole runtime available in WebAssembly, but I suspect that shouldn't be the focus of the PowerShell team.  Nor do I want it to be my focus.  Until there's a very clear and easy way to compile all of .NET core and PowerShell into WebAssembly, I see a lot of complexity and a fairly limited "win" with a number of security implications.

That stated, I do see a few valuable scenarios with respect to PowerShell and WebAssembly.  These are scenarios I will seek to support with PipeScript.

1. [ ] Being able to manipulate WebAssembly (.wat) files.  WAT generation is hard, and PowerShell could help make it easier.  The basics of this are being tracked in [this issue in PipeScript](https://github.com/StartAutomating/PipeScript/issues/239).
2. [ ] Being able to perform limited translation to .WAT.  A sufficiently simple [ScriptBlock] should be translatable to WAT.  This is a bigger can of worms, as it relates to how one would generally translate PowerShell into other languages.  WAT may be a special challenge because of it's very primal format, in the same sort of way that translating PowerShell to actual assembly would be.
3. [ ] Higher level languages than assemblers should be easier to translate tool, and their toolchains should be easier to connect.  For instance, if one creates the capability to translate from PowerShell to C, one would get WASM support essentially "for free".

In closing:

This may be cool and may be interesting, but it's a long road to snapping our fingers and making it happen.  You're more than welcome to help walk it.

**Reply by Ayanmullick** [reply]
Thank you for sharing your thoughts. I thought it was possible to compile .Net to WebAssembly.

[Is it possible to compile .NET class library to WebAssembly to call from JavaScript? : dotnet (reddit.com)](https://www.reddit.com/r/dotnet/comments/ttx4xq/comment/i30it4y/?utm_source=share&utm_medium=web2x&context=3)

I haven't tried hands-on though.

**Reply by fMichaleczek** [reply]
To compile PowerShell to low level WASM, you will have to disable generic, reflection and dynamic. It should be possible with the V2 source code before the DLR but i dont think it's possible now. At this moment, even if PowerShell was compilable to Native AOT (Ahead of Time), I don't know if the dotnet toolchain is able to produce low level wasm outside experimental.
Lambda expressions are not considered like Reflection or Dynamic and are able to produce object in very restricted host (an iphone).

**Comment by fMichaleczek** [comment]
![image](https://user-images.githubusercontent.com/6187529/193403052-2e8f3f67-4a00-49bf-b848-bc38bf9e2022.png)
https://twitter.com/_Flavien/status/1254569119560671233?s=20&t=gOudPvzJLACJ-jIVgA135Q

[PowerShell WASM Demo](https://fmichaleczek.github.io/PowerShellLab/)
![image](https://user-images.githubusercontent.com/6187529/193403121-4742c408-fb33-40e6-83c3-c2b2ac7cc6f6.png)
There is no formatter, so you have to add Out-String

Blazor is one thread model, so PowerShell can't await task and can't use the WebClient API.
It's a big problem because it means PowerShell can't escape from the browser sandbox.

![image](https://user-images.githubusercontent.com/6187529/193403504-23b93ce6-28ae-4de7-ad24-3a2724e81c7f.png)

The dotnet team announced that multithread will be on dotnet 8.0
![image](https://user-images.githubusercontent.com/6187529/193403290-691a2402-300a-43c8-adb0-a8362b81227b.png)

**Reply by dimitertodorov** [reply]
Could you share the code for this? I am looking at running some PowerShell in a WASM app I have in mind.

**Reply by fMichaleczek** [reply]
> Could you share the code for this? I am looking at running some PowerShell in a WASM app I have in mind

dotnet 8.0 will offer a lot of low-level API and it will improve the PowerShell integration. (More about to develop a simple javascript layer to improve PowerShell communication).

I will send an update soon

**Comment by kilasuit** [comment]
Could you edit the title to be `PowerShell in WebAssembly` as there's no real need for the `Make powershell v9 great again` in the title as PowerShell is already a great language, it can however only become better in future.

**Reply by kilasuit** [reply]
@kasini3000 it may not have come across correctly due to language differences but I had asked you to edit the title in the above comment. So please will you edit this to be `PowerShell in WebAssembly` so that we can have better tracking going forward please

**Comment by rbleattler** [comment]
I'll be keeping my eye on this conversation and looking to contribute should things progress. 

I also want to note that in my opinion it is unhelpful to the community to state that that "PowerShell is not general purpose" or is just "really good at automating tasks". I certainly can understand how one might develop that opinion, but I have seen how powerful PowerShell can be, and I have built some very complex projects around it. Just look at the incredible products @adamdriscoll and his team have built around it. 

I hope to see this idea gain traction. Thanks to those who shared links to projects working on this.

**Reply by rhubarb-geek-nz** [reply]
> I also want to note that in my opinion it is unhelpful to the community to state that that 

```
$ dpkg -l powershell
ii  powershell     7.3.2-1.deb  amd64        PowerShell is an automation and configuration management platform.
```

> I certainly can understand how one might develop that opinion

**Reply by StartAutomating** [reply]
Yeah.  I think one of the ironies of PowerShell is that while it is a great scripting language, it generally doesn't "lead" with that in a number of forums.

From [What Is PowerShell on learn.microsoft.com](https://learn.microsoft.com/en-us/powershell/scripting/overview?view=powershell-7.3)
> PowerShell is a cross-platform task automation solution made up of a command-line shell, a scripting language, and a configuration management framework. PowerShell runs on Windows, Linux, and macOS.

From [PowerShell's GitHub README]:
> [PowerShell](https://learn.microsoft.com/powershell/scripting/overview) is a cross-platform (Windows, Linux, and macOS) automation and  configuration tool/framework that works well with your existing tools and is optimized for dealing with structured data (e.g. JSON, CSV, XML, etc.), REST APIs, and object models. It includes a command-line shell, an associated scripting language and a framework for processing cmdlets.

I'm trying to do my part to rectify this rhetoric.

From the [PowerShell Guide](https://powershellguide.com/PowerShell/)
> PowerShell is a [scripting language](https://powershellguide.com/Languages/Scripting-Languages) and a [shell](https://powershellguide.com/Shells), and a rather good one of each of these things.

IMO, It would be nice if PowerShell embraced "Scripting Language" much more directly in it's literature, not as a footnote.

**Reply by rbleattler** [reply]
@rhubarb-geek-nz as @StartAutomating points out in the reply above, the rhetoric doesn't reflect the reality. PowerShell is an excellent scripting language. Is it as mature as Python? No, but it's also only half Python's age and has undergone some massive changes in its time. Personally I find PowerShell to be the most robust out of the box solution, with the lowest barrier to entry -- especially for anyone familiar with the dotnet space.

**Reply by rhubarb-geek-nz** [reply]
Yes, I think that Python is a good comparison but I don't think that makes them equivalent. I find it is certainly the best way of scripting on Windows especially given the base PowerShell is there out of the box.

**Comment by rhubarb-geek-nz** [comment]
Can somebody summarise the mechanism of how this works? My understanding is that WebAssembly is based on a byte code engine in the web browser. So to run any PowerShell you effectively need the dotnet CLR VM engine and any native modules compiled to web assembly, then that can interpret the rest of the dotnet class libraries and PowerShell itself when running in the WebBrowser. As this is a scripting language you can't predict what is required so you need to download of the order of 60M of binaries to the client browser in order to run any PowerShell script. Is this how it works?

As a developer and user of Azure I completely understand the use case of running PowerShell in a browser as a console application in order to not require an Azure VM to run it, it would still need Azure Storage for scripts and environment. So I can envisage all of PowerShell, dotnet runtime and Azure tools being downloaded to the client for local execution within the browser.

What I am less clear on is the expectation of being able to put PowerShell script fragments directly into a page of HTML for execution in the within the browser. This would at least imply the downloading of a PowerShell interpreter of some kind to run in the browser. Is this the full interpreter or some special cutdown PowerShell interpreter that only downloads CLR modules and cmdlets as they are needed? Are they cached in the browser between pages and shared between sites?

What makes this more likely to succeed than Java Applets, Microsoft Silverlight and Adobe Flash?

**Reply by fMichaleczek** [reply]
The Web a no concept of a Terminal. A Console exists in Chromium but in the debug tools. Write-Host will display message here (Console.WriteLine API is mapped there) . Having PowerShell inside a Browser means extensions, debug tools and webapp. A minimal  Terminal webapp could be technically possible but there is no standard for ANSI. The available JS Terminal are mostly limited. (no VT100 is missing).

To start, PowerShell script could be embedded in html like ```<powershell-script>echo 'test' </powershell-script>``` throught a minimal vanilla javascript framework.

I can make a port of the language but not the shell part. I'm sorry you imagine to do the same things but it's more about the root language then the daily admin life. I mean PowerShell is a language with no name. 

WASM is divised into 2 parts : Browser et Process. The first one is limited to a Browser Sandbox and the second is inside a node.js process with more API to escape. Browser is the subject that leads the WASM technology into dotnet. There is a secund time where powershell.wasm will run into a WASM app inside a WASM container into an Kubernetes pode. 

The last dotnet WASM Browser has multithread experimental support. PowerShell need 2 threads to handle an async web request. The previous PowerShell port wasn't able to escape the Browser VM and the next spring will be to implement and test threading and await an async method (to download a web file from a trust url).

Download size should be around 20-30MB. It depends on the trimming (Only keep minimal dotnet platform assemblies, applicable to Web) 

PS : Android is the same problem. No console. No terminal. 
There is also hybrid WASM.
![Blazor Hybrid](https://devblogs.microsoft.com/dotnet/wp-content/uploads/sites/16/2020/07/blazor-hybrid-ui.png)

**Reply by rhubarb-geek-nz** [reply]
Thanks for that detailed explanation. My concept of terminal was not the debugging console in a browser, but the full console that one sees in the Azure Portal when running bash or PowerShell session within the browser. PowerShell is running on a VM somewhere in Azure with the presentation appearing in the browser.

What are the advantages of running PowerShell on WASM in Kubernetes given we already have native AMD64 and ARM64 implementations running in Docker?

**Reply by fMichaleczek** [reply]
Well, the Chromium model about diagnostics is big. And you will be able also to enter pwsh('echo test'). 

Technologies stacking : 
![Uno Wasm Debugger in Chrome Tools](https://platform.uno/docs/articles/Assets/quick-start/wasm-debugger-step-02.png) 

PowerShell has to write html, css and communicate with javascript framework. And it will not be a better than existant one. But script guys will be able to write frontend webpage. 
And HTML5 bring a lot of interesting feature (mount an USB drive, take a picture, etc..). 

Yes, there is another part, communicate with a custom server, but it's another big step and the project will be limited to a very compatible secure endpoint (see SharedArray and Meltdown on mobile) 

WASM is a kind of cloud glibc where all api are sandboxed. Imagine a tab as a container in chrome. The sandbox API should become a standard for web and non-web. 
 ![](https://platform.uno/docs/articles/Assets/quick-start/wasm-debugger-step-02.png)

**Reply by fMichaleczek** [reply]
This Blazor Console PoC is a beginning but it's very slow without AOT (compile to native) and PowerShell is incompatible with AOT. 

I hope using directly the dotnet low level API will help to be more light than Blazor. 

 ![](https://user-images.githubusercontent.com/3766240/163650530-01c3cbf2-d691-4de5-9d47-4771fb886448.png) 
https://github.com/Blazor-Console/HACC.Demo

**Reply by fMichaleczek** [reply]
WebCIL file format creation to replace DLL in WebContext. (dotnet 8.0 development) 
https://github.com/dotnet/runtime/pull/79416

**Reply by rhubarb-geek-nz** [reply]
> WASM is a kind of cloud glibc where all api are sandboxed. 

My understanding is that it is its own assembly language based on an abstract machine along with an API for applications. So PowerShell uses CLR and runs on dotnet which is then compiled to the host. If the host is WASM then that again converts it to the actual machine it is running on.

**Reply by fMichaleczek** [reply]
PeachPie is a php port for aspnet.core

A student had WASM support with Blazor :  
https://www.peachpie.io/2021/08/run-debug-php-browser.html

The Thesis
https://github.com/TomatorCZ/Bachelor_thesis/blob/main/docs/thesis/thesis.pdf

PS : Peachpie is compiled like C#. The model is different than PowerShell which is dynamic.

**Reply by fMichaleczek** [reply]
>  So PowerShell uses CLR.

No, a special MonoEngine, who was a fork of MoonLight, the open source version of SilverLight. But the BCL are shared between CoreCLR, MonoEngine and MonoEngine WASM.

So my goal is use MonoEngine (Android) and MonoEngine WASM. Apple refuse Dynamic Code so it's dead.

**Comment by rhubarb-geek-nz** [comment]
One limitation of PowerShell in modern user interface code is it is thread based and doesn't currently support the async/await/Task model. One typically uses await in Javascript when calling back to a server. Hopefully this is on the roadmap.

**Reply by fMichaleczek** [reply]
no hope. I don't see how one could solve what never was. Models with a shared thread and synchronized contexts must be accepted as advanced programming. PowerShell can do a lot of things but being dynamic and compiled at the same time is not possible. We cannot pre-program a state machine to manage an asynchronous processing tree, but we can imagine a limited promise system. The prerequisite is at least 2 threads. We will have to get used to sharing with a synchronized UI context (Winform, WPF, Android, Browser) and each having its own way of doing things. We will see what MAUI and Uno will do on the subject, in the meantime we have to go down to a low level on WASM, which these generic frameworks will not do.

**Comment by rhubarb-geek-nz** [comment]
What is the actual goal is here? Is it something has to be done with WebAssembly or we want PowerShell developers to be able to write web-pages? It looks like WebAssembly here is a solution looking for a problem. 

If you want to use PowerShell to produce web-pages there is nothing stopping you, you can write a web-server in PowerShell right now. 

If you want to embed PowerShell in HTML like Java Script then would it not be simpler to have translators that convert your embedded PowerShell into JavaScript in a similar manner to TypeScript? That way JavaScript is still served to the client and no issues with iPhone.

**Reply by fMichaleczek** [reply]
If we follow your mindset, why not use bash instead of PowerShell ?
I understand your frustration but I can't do much about it. 😕

Anyway, the topic of the thread is clear about the main subject : Just watch the PyScript demos from the first message :  
https://pyscript.net/examples/

It's about Hello world, Clock demo, REPL or Todo App, nothing more fancy.

**Reply by rhubarb-geek-nz** [reply]
> If we follow your mindset, why not use bash instead of PowerShell ? 

Bash? We hash-bang with /bin/sh here!

Actually, my cross-platform scripting language of choice is PowerShell, because it is a far better fit on Linux and macOS than [MinGW](https://www.mingw-w64.org) or similar is on Windows.

> It's about Hello world, Clock demo, REPL or Todo App, nothing more fancy.

I wish you well, it is a tall stack to climb, dotnet first then PowerShell. I notice in the demos that Python supports async IO with await and they happily worked on an iPhone.

**Reply by fMichaleczek** [reply]
Yes but PowerShell language never came out of its shell. And its users have a different IT culture from developers. If we extract the language to do something else, mentalities must also evolve. A language is not a technology. If I put PowerShell on a bike. At best, I could pedal. And if people ask if we can start a server with it, it's going to be a long time before we do the first mile with this bike. You have to focus on what it does, not on what it no longer does or will never do.

**Reply by rhubarb-geek-nz** [reply]
Continuing with the comparison with Python, CPython is a popular reference implementation of the language but there is nothing stopping you using a different Python interpreter. You could support the PowerShell language itself, if it truly exists as a language, using either a completely different interpreter or a different compiler. Traditionally with languages you can separate the language from the commonly used libraries, eg on Windows a C/C++ developer would use the Win32 API in preference to the original C library that originated on UNIX . I would imagine that web development would use only a small subset of existing Cmdlets. So imagine a PowerShell language interpreter that was not based on dotnet, and perhaps written directly in C, Rust or Go using only the WebAssembly library.

**Reply by rbleattler** [reply]
> Yes but PowerShell language never came out of its shell. And its users have a different IT culture from developers. If we extract the language to do something else, mentalities must also evolve. A language is not a technology. If I put PowerShell on a bike. At best, I could pedal. And if people ask if we can start a server with it, it's going to be a long time before we do the first mile with this bike. You have to focus on what it does, not on what it no longer does or will never do.

"Its users"? I am among its users. I work in a DevOps capacity and spend the overwhelming majority of my time working to best understand the needs and streamline the work of the developers I support. This is so narrowminded and fails to acknowledge the potential of this language. It "never came out of its shell" is putting on blinders and refusing to acknowledge all the incredible work people are doing with it -- without any real support to expand it. 


> Continuing with the comparison with Python, CPython is a popular reference implementation of the language but there is nothing stopping you using a different Python interpreter. You could support the PowerShell language itself, if it truly exists as a language, using either a completely different interpreter or a different compiler. Traditionally with languages you can separate the language from the commonly used libraries, eg on Windows a C/C++ developer would use the Win32 API in preference to the original C library that originated on UNIX . I would imagine that web development would use only a small subset of existing Cmdlets. So imagine a PowerShell language interpreter that was not based on dotnet, and perhaps written directly in C, Rust or Go using only the WebAssembly library.

The source is open -- there is absolutely nothing stopping anyone from building an interpreter using C, Rust, Go, or anything else. If someone wanted to do so, the best place to start would likely be with the [parser](https://github.com/PowerShell/PowerShell/blob/037e12eddc850393d2538a26abad4a67e048a989/src/System.Management.Automation/engine/parser/Parser.cs)...
